Subscriptions to Salesforce

Alexalex1432 opened the pull request:
Adding the ability to accept subscriptions from Funraise, link opportunities & contacts to them

Resolving FUN-7068

beta install link here: https://login.salesforce.com/packaging/installPackage.apexp?p0=04t1C000000OD3C

Commit:
Adding the ability to accept subscriptions from Funraise, link opportunities & contacts to them

Resolving FUN-7068

funraise-brian commented on src/classes/frDonation.cls:
```diff
@@ -98,11 +99,22 @@ public class frDonation extends frModel {
+        if (String.isNotBlank(subscriptionId)) {
+            List<Subscription__c> subscriptions = [SELECT Id FROM Subscription__c WHERE fr_ID__c = :subscriptionId];
+            if (subscriptions.size() > 0) {
+                o.Subscription__c = subscriptions.get(0).Id;
```
Hm seems slightly inconsistent when comparing line 103 with 113. 103 appears to be an array while 113 is a List

Alexalex1432 replied:
Converted the other one to a list.  I think they are functionally equivalent, and I suppose the List<> is more visually pleasing to me

funraise-brian commented on src/classes/frSubscription.cls:
```diff
@@ -0,0 +1,119 @@
+        } catch (DMLException e) {
+            insert new Error__c(Error__c =
+                                'Failed to insert funraise subscription id ' + funraiseId +
+                                '. Exception: ' + e.getMessage());
```
Is the exception going to include the field names? seems like it would be helpful if we were going to troubleshoot any problems that may occur with the `funraiseFieldName` -> `salesforceFieldName` conversion code

Alexalex1432 replied:
Good catch.  I just checked to see what would happen if we try to call `sObject.put('invalidField', value)`

It throws a runtime exception, which would be unhandled and would trigger an email off to the managed package contact (I think that's me).  I think that would be the correct behavior as we are not expecting any of the fields to not match here

funraise-brian commented on src/classes/frSubscription.cls:
```diff
@@ -0,0 +1,119 @@
+            String salesforceFieldName = funraiseFieldName.replaceAll(regex, replacement);
+            salesforceFieldName =+ 'funraise__' + salesforceFieldName.capitalize() + '__c';
+            Schema.SObjectField field = fields.get(salesforceFieldName);
+            frUtil.write(subscription, field, funraiseFieldName, request.get(funraiseFieldName), funraiseId);
```
Any special handling/conversion that needs to happen with the difference in types between `lastPaymentDate` and `nextPaymentDate`?

Alexalex1432 replied:
Not needed since we've removed lastPaymentDate as a field now

Commit:
Make donation & donor test classes consistent with the other classes. Fix case problem on @isTest annotation in frSetupControllerTest

frWSDonationControllerTest -> frDonationTest
frWSDonorControllerTest -> frDonorTest (both of these already existed so I just combined the classes, also removed some tests from frDonorTest as I'm refactoring that class and some of the methods have been removed)
frWSDonorEmailControllerTest -> frDonorEmailsTest (also a merge since they both existed)

unrelated to the feature in question but addressing some tech debt as part of resolving FUN-7068

Commit:
Some refactoring and simplification of frDonation & frDonor.  These were the first classes in this project and isn't really consistent with how we're writing them now.

Also make it so that if we find a match with an existing contact, we populate the funraise id and save that immediately so that if any mapping errors prevent the record from saving, it still has the funraise id on it for future integration attempts.  It also allows us to remove the update/upsert check that was happening, which led to sometimes hitting duplicate value exceptions in the fr_ID__c field.  Now we will only do upserts on the fr_ID__c.

Also addressed some tech debt where for Campaign supporterId and parentGoalId were primitive long instead of Longs, so we had to set them to 0 instead of null when there wasn't a value.  That's fixed on the platform side so we don't have to check for 0 anymore

Commit:
Removing datapoints that are derived from other data.  On the salesforce side both of these values can be calculated by checking the number of opportunities associated with a subscription and the most recent creation date.

Resolving FUN-7068

Alexalex1432 commented:
Managed package beta with the feedback included from the most recent commits
https://login.salesforce.com/packaging/installPackage.apexp?p0=04t1C000000OD3W
